### PR TITLE
make checking the julia executable more robust

### PR DESCRIPTION
**Pull request by pfitzseb**
and spawn julia in the first open project folder (or its parent if that is a file), as per [here](https://github.com/JunoLab/atom-julia-client/issues/124#issuecomment-193303789).

### CODE REVIEW

**Comment by pfitzseb** [comment]
@MikeInnes Actually, we might get away with deleting the fallback entirely -- I think this implementation here is quite comprehensive. What do you think?

File: lib/connection/process.coffee [hunk 1]
```diff
@@ -36,28 +38,49 @@ module.exports = jlprocess =
...
+  # spawn Julia in the first open project folder (or its parent folder for files)
   workingDir: ->
     paths = atom.workspace.project.getDirectories()
-    if paths.length == 1 and fs.statSync(paths[0].path).isDirectory()
-      paths[0].path
+    if fs.statSync(paths[0].path).isFile()
+      path.dirname paths[0].path
     else
-      process.env.HOME || process.env.USERPROFILE
+      paths[0].path
```
Comment (MikeInnes) [inline lib/connection/process.coffee]: I think it's possible for their to be 0 project dirs, right?

File: lib/connection/process.coffee [hunk 2]
```diff
@@ -36,28 +38,49 @@ module.exports = jlprocess =
...
     if p == '[bundle]' then p = @bundledExe()
     p
 
-  checkPath: (path) ->
-    new Promise (resolve) ->
-      fs.exists path, (exists) ->
+  checkPath: (p) ->
+    new Promise (resolve, reject) =>
+      # check whether path exists
+      fs.exists p, (exists) =>
```
Comment (MikeInnes) [inline lib/connection/process.coffee]: We might be able to flatten this big chain a bit / reduce the filesystem calls by using `fs.stat` to both check for existence and file-ness.

File: lib/connection/process.coffee (hunk 1 above)
Comment (pfitzseb) [inline lib/connection/process.coffee]: True, didn't think of that. Should be fixed now.

File: lib/connection/process.coffee [hunk 3]
```diff
@@ -73,12 +100,8 @@ module.exports = jlprocess =
...
     client.booting()
 
-    @checkPath(@jlpath()).then (exists) =>
-      if not exists
-        @jlNotFound @jlpath()
-        client.cancelBoot()
-        return
-
+    @checkPath(@jlpath())
+    .then =>
```
Comment (MikeInnes) [inline lib/connection/process.coffee]: We usually indent the `.then` and `.catch` to make it clear it's not a new statement

File: lib/connection/process.coffee (hunk 2 above)
Comment (pfitzseb) [inline lib/connection/process.coffee]: Done. If we were to remove the `which` fallback it'd look even nicer ;)

File: lib/connection/process.coffee (hunk 3 above)
Comment (pfitzseb) [inline lib/connection/process.coffee]: Also done.

**Comment by MikeInnes** [comment]
Great changeset. I'm wondering about the exe-in-folder check though – is it really that common to input the folder just above the julia exe? It seems like there are a lot of small user mistakes like that that we could correct for, but at some point we have to give up and write better docs :)

**Comment by pfitzseb** [comment]
Thanks for the review, @MikeInnes!

The exe-in-folder thingy came up at least twice, I think, so it seemed logical to fix it... Better docs would of course be nice, even though the setting explicitly says that we want the *binary*.

**Comment by pfitzseb** [comment]
Alright, I've refactored stuff a bit -- if we want to keep the automatic checking for the exe then this should be good to go, otherwise I'll rip it out ;)

**Comment by MikeInnes** [comment]
This looks good. I was vaguely worried that if we supported setting the path as a folder, users would rely on it and we'd never be able to change it back; but since you're actually resetting the path config to be correct that's actually not an issue.

Thanks for the patch!

**Pull request merged**